File: 98-validate-binary-search-tree/validate-binary-search-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
bool isValidBST(struct TreeNode* root) {
    int* a = NULL;
    int n;
    bool helper(struct TreeNode * root) {
        if (!root)
            return true;
        if (!helper(root->left))
            return false;
        if (a && *a >= root->val)
            return false;
        n = root->val;
        a = &n;
        if (!helper(root->right))
            return false;
        return true;
    }
    return helper(root);
}

File: 98-validate-binary-search-tree/validate-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    a=float('-inf')
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return True
        if not self.isValidBST(root.left):
            return False
        if self.a>=root.val:
            return False
        self.a=root.val
        if not self.isValidBST(root.right):
            return False
        return True

        